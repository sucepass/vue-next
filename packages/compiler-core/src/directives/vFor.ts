import { createDirectiveTransform } from '../transform'
import { NodeTypes, ExpressionNode } from '../ast'
import { createCompilerError, ErrorCodes } from '../errors'
import { getInnerRange } from '../utils'

const forAliasRE = /([\s\S]*?)(?:(?<=\))|\s+)(?:in|of)\s+([\s\S]*)/
const forIteratorRE = /,([^,\}\]]*)(?:,([^,\}\]]*))?$/
const stripParensRE = /^\(|\)$/g

export const transformFor = createDirectiveTransform(
  'for',
  (node, dir, context) => {
    if (dir.exp) {
      const aliases = parseAliasExpressions(dir.exp.content)

      if (aliases) {
        context.replaceNode({
          type: NodeTypes.FOR,
          loc: node.loc,
          source: maybeCreateExpression(
            aliases.source,
            dir.exp
          ) as ExpressionNode,
          valueAlias: maybeCreateExpression(aliases.value, dir.exp),
          keyAlias: maybeCreateExpression(aliases.key, dir.exp),
          objectIndexAlias: maybeCreateExpression(aliases.index, dir.exp),
          children: [node]
        })
      } else {
        context.onError(
          createCompilerError(
            ErrorCodes.X_FOR_MALFORMED_EXPRESSION,
            dir.loc.start
          )
        )
      }
    } else {
      context.onError(
        createCompilerError(ErrorCodes.X_FOR_NO_EXPRESSION, dir.loc.start)
      )
    }
  }
)

interface AliasExpression {
  offset: number
  content: string
}

interface AliasExpressions {
  source: AliasExpression
  value: AliasExpression | undefined
  key: AliasExpression | undefined
  index: AliasExpression | undefined
}

function parseAliasExpressions(source: string): AliasExpressions | null {
  const inMatch = source.match(forAliasRE)
  if (!inMatch) return null

  const [, LHS, RHS] = inMatch
  const result: AliasExpressions = {
    source: {
      offset: source.indexOf(RHS, LHS.length),
      content: RHS.trim()
    },
    value: undefined,
    key: undefined,
    index: undefined
  }

  let valueContent = LHS.trim()
    .replace(stripParensRE, '')
    .trim()
  const trimmedOffset = LHS.indexOf(valueContent)

  const iteratorMatch = valueContent.match(forIteratorRE)
  if (iteratorMatch) {
    valueContent = valueContent.replace(forIteratorRE, '').trim()

    const keyContent = iteratorMatch[1].trim()
    if (keyContent) {
      result.key = {
        offset: source.indexOf(keyContent, trimmedOffset + valueContent.length),
        content: keyContent
      }
    }

    if (iteratorMatch[2]) {
      const indexContent = iteratorMatch[2].trim()

      if (indexContent) {
        result.index = {
          offset: source.indexOf(
            indexContent,
            result.key
              ? result.key.offset + result.key.content.length
              : trimmedOffset + valueContent.length
          ),
          content: indexContent
        }
      }
    }
  }

  if (valueContent) {
    result.value = {
      offset: trimmedOffset,
      content: valueContent
    }
  }

  return result
}

function maybeCreateExpression(
  alias: AliasExpression | undefined,
  node: ExpressionNode
): ExpressionNode | undefined {
  if (alias) {
    return {
      type: NodeTypes.EXPRESSION,
      loc: getInnerRange(node.loc, alias.offset, alias.content.length),
      content: alias.content,
      isStatic: false
    }
  }
}
