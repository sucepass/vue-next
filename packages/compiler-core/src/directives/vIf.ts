import { createDirectiveTransform } from '../transform'
import {
  NodeTypes,
  ElementTypes,
  ElementNode,
  DirectiveNode,
  IfBranchNode
} from '../ast'
import { createCompilerError, ErrorCodes } from '../errors'

export const transformIf = createDirectiveTransform(
  /^(if|else|else-if)$/,
  (node, dir, context) => {
    if (dir.name === 'if') {
      context.replaceNode({
        type: NodeTypes.IF,
        loc: node.loc,
        branches: [createIfBranch(node, dir)]
      })
    } else {
      // locate the adjacent v-if
      const siblings = context.parent.children
      const comments = []
      let i = siblings.indexOf(node)
      while (i-- >= -1) {
        const sibling = siblings[i]
        if (__DEV__ && sibling && sibling.type === NodeTypes.COMMENT) {
          context.removeNode(sibling)
          comments.unshift(sibling)
          continue
        }
        if (sibling && sibling.type === NodeTypes.IF) {
          // move the node to the if node's branches
          context.removeNode()
          const branch = createIfBranch(node, dir)
          if (__DEV__ && comments.length) {
            branch.children = [...comments, ...branch.children]
          }
          sibling.branches.push(branch)
        } else {
          context.onError(
            createCompilerError(
              dir.name === 'else'
                ? ErrorCodes.X_ELSE_NO_ADJACENT_IF
                : ErrorCodes.X_ELSE_IF_NO_ADJACENT_IF,
              node.loc.start
            )
          )
        }
        break
      }
    }
  }
)

function createIfBranch(node: ElementNode, dir: DirectiveNode): IfBranchNode {
  return {
    type: NodeTypes.IF_BRANCH,
    loc: node.loc,
    condition: dir.name === 'else' ? undefined : dir.exp,
    children: node.tagType === ElementTypes.TEMPLATE ? node.children : [node]
  }
}
