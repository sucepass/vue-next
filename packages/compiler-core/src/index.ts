export { baseCompile } from './compile'

// Also expose lower level APIs & types
export {
  CompilerOptions,
  ParserOptions,
  TransformOptions,
  CodegenOptions
} from './options'
export { baseParse, TextModes } from './parse'
export {
  transform,
  createStructuralDirectiveTransform,
  TransformContext,
  NodeTransform,
  StructuralDirectiveTransform,
  DirectiveTransform
} from './transform'
export { generate, CodegenContext, CodegenResult } from './codegen'
export {
  ErrorCodes,
  CoreCompilerError,
  CompilerError,
  createCompilerError
} from './errors'
export * from './ast'
export * from './utils'
export { registerRuntimeHelpers } from './runtimeHelpers'

// expose transforms so higher-order compilers can import and extend them
export { transformModel } from './transforms/vModel'
export { transformOn } from './transforms/vOn'

// utility, but need to rewrite typing to avoid dts relying on @vue/shared
import { generateCodeFrame as _genCodeFrame } from '@vue/shared'
const generateCodeFrame = _genCodeFrame as (
  source: string,
  start?: number,
  end?: number
) => string
export { generateCodeFrame }
