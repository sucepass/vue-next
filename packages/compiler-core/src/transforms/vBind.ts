import { DirectiveTransform } from '../transform'
import { createObjectProperty, createExpression } from '../ast'
import { createCompilerError, ErrorCodes } from '../errors'
import { camelize } from '@vue/shared'

// v-bind without arg is handled directly in ./element.ts due to it affecting
// codegen for the entire props object. This transform here is only for v-bind
// *with* args.
export const transformBind: DirectiveTransform = (
  { exp, arg, modifiers, loc },
  context
) => {
  if (!exp) {
    context.onError(createCompilerError(ErrorCodes.X_V_BIND_NO_EXPRESSION, loc))
  }
  // .prop is no longer necessary due to new patch behavior
  // .sync is replced by v-model:arg
  if (modifiers.includes('camel')) {
    arg!.content = camelize(arg!.content)
  }
  return {
    props: createObjectProperty(
      arg!,
      exp || createExpression('', true, loc),
      loc
    ),
    needRuntime: false
  }
}
