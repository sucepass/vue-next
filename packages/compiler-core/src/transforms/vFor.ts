import {
  createStructuralDirectiveTransform,
  TransformContext
} from '../transform'
import {
  NodeTypes,
  ExpressionNode,
  createExpression,
  SourceLocation
} from '../ast'
import { createCompilerError, ErrorCodes } from '../errors'
import { getInnerRange } from '../utils'
import { RENDER_LIST } from '../runtimeConstants'
import { processExpression } from './transformExpression'

export const transformFor = createStructuralDirectiveTransform(
  'for',
  (node, dir, context) => {
    if (dir.exp) {
      const parseResult = parseForExpression(dir.exp, context)

      if (parseResult) {
        context.helper(RENDER_LIST)
        const { source, value, key, index } = parseResult

        context.replaceNode({
          type: NodeTypes.FOR,
          loc: dir.loc,
          source,
          valueAlias: value,
          keyAlias: key,
          objectIndexAlias: index,
          children: [node]
        })

        // scope management
        const { addIdentifier, removeIdentifier } = context

        // inject identifiers to context
        value && addIdentifier(value)
        key && addIdentifier(key)
        index && addIdentifier(index)

        return () => {
          // remove injected identifiers on exit
          value && removeIdentifier(value)
          key && removeIdentifier(key)
          index && removeIdentifier(index)
        }
      } else {
        context.onError(
          createCompilerError(ErrorCodes.X_FOR_MALFORMED_EXPRESSION, dir.loc)
        )
      }
    } else {
      context.onError(
        createCompilerError(ErrorCodes.X_FOR_NO_EXPRESSION, dir.loc)
      )
    }
  }
)

const forAliasRE = /([\s\S]*?)\s+(?:in|of)\s+([\s\S]*)/
const forIteratorRE = /,([^,\}\]]*)(?:,([^,\}\]]*))?$/
const stripParensRE = /^\(|\)$/g

interface ForParseResult {
  source: ExpressionNode
  value: ExpressionNode | undefined
  key: ExpressionNode | undefined
  index: ExpressionNode | undefined
}

function parseForExpression(
  input: ExpressionNode,
  context: TransformContext
): ForParseResult | null {
  const loc = input.loc
  const source = input.content
  const inMatch = source.match(forAliasRE)
  if (!inMatch) return null

  const [, LHS, RHS] = inMatch
  const result: ForParseResult = {
    source: createAliasExpression(
      loc,
      RHS.trim(),
      source.indexOf(RHS, LHS.length),
      context,
      context.prefixIdentifiers
    ),
    value: undefined,
    key: undefined,
    index: undefined
  }

  let valueContent = LHS.trim()
    .replace(stripParensRE, '')
    .trim()
  const trimmedOffset = LHS.indexOf(valueContent)

  const iteratorMatch = valueContent.match(forIteratorRE)
  if (iteratorMatch) {
    valueContent = valueContent.replace(forIteratorRE, '').trim()

    const keyContent = iteratorMatch[1].trim()
    let keyOffset: number | undefined
    if (keyContent) {
      keyOffset = source.indexOf(
        keyContent,
        trimmedOffset + valueContent.length
      )
      result.key = createAliasExpression(loc, keyContent, keyOffset, context)
    }

    if (iteratorMatch[2]) {
      const indexContent = iteratorMatch[2].trim()

      if (indexContent) {
        result.index = createAliasExpression(
          loc,
          indexContent,
          source.indexOf(
            indexContent,
            result.key
              ? keyOffset! + keyContent.length
              : trimmedOffset + valueContent.length
          ),
          context
        )
      }
    }
  }

  if (valueContent) {
    result.value = createAliasExpression(
      loc,
      valueContent,
      trimmedOffset,
      context
    )
  }

  return result
}

function createAliasExpression(
  range: SourceLocation,
  content: string,
  offset: number,
  context: TransformContext,
  process: boolean = false
): ExpressionNode {
  const exp = createExpression(
    content,
    false,
    getInnerRange(range, offset, content.length)
  )
  if (!__BROWSER__ && process) {
    processExpression(exp, context)
  }
  return exp
}
