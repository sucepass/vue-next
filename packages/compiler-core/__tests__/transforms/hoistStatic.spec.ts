import { parse, transform, NodeTypes, generate } from '../../src'
import {
  OPEN_BLOCK,
  CREATE_BLOCK,
  CREATE_VNODE,
  APPLY_DIRECTIVES,
  FRAGMENT,
  RENDER_LIST
} from '../../src/runtimeHelpers'
import { transformElement } from '../../src/transforms/transformElement'
import { transformIf } from '../../src/transforms/vIf'
import { transformFor } from '../../src/transforms/vFor'
import { transformBind } from '../../src/transforms/vBind'
import { createObjectMatcher, genFlagText } from '../testUtils'
import { PatchFlags } from '@vue/shared'

function transformWithHoist(template: string) {
  const ast = parse(template)
  transform(ast, {
    hoistStatic: true,
    nodeTransforms: [transformIf, transformFor, transformElement],
    directiveTransforms: {
      bind: transformBind
    }
  })
  expect(ast.codegenNode).toMatchObject({
    type: NodeTypes.JS_SEQUENCE_EXPRESSION,
    expressions: [
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: OPEN_BLOCK
      },
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_BLOCK
      }
    ]
  })
  return {
    root: ast,
    args: (ast.codegenNode as any).expressions[1].arguments
  }
}

describe('compiler: hoistStatic transform', () => {
  test('should NOT hoist root node', () => {
    // if the whole tree is static, the root still needs to be a block
    // so that it's patched in optimized mode to skip children
    const { root, args } = transformWithHoist(`<div/>`)
    expect(root.hoists.length).toBe(0)
    expect(args).toEqual([`"div"`])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist simple element', () => {
    const { root, args } = transformWithHoist(
      `<div><span class="inline">hello</span></div>`
    )
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [
          `"span"`,
          createObjectMatcher({ class: 'inline' }),
          {
            type: NodeTypes.TEXT,
            content: `hello`
          }
        ]
      }
    ])
    expect(args).toMatchObject([
      `"div"`,
      `null`,
      [
        {
          type: NodeTypes.ELEMENT,
          codegenNode: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `_hoisted_1`
          }
        }
      ]
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist nested static tree', () => {
    const { root, args } = transformWithHoist(
      `<div><p><span/><span/></p></div>`
    )
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [
          `"p"`,
          `null`,
          [
            { type: NodeTypes.ELEMENT, tag: `span` },
            { type: NodeTypes.ELEMENT, tag: `span` }
          ]
        ]
      }
    ])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_1`
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist nested static tree with comments', () => {
    const { root, args } = transformWithHoist(
      `<div><div><!--comment--></div></div>`
    )
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [
          `"div"`,
          `null`,
          [{ type: NodeTypes.COMMENT, content: `comment` }]
        ]
      }
    ])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_1`
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist siblings with common non-hoistable parent', () => {
    const { root, args } = transformWithHoist(`<div><span/><div/></div>`)
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [`"span"`]
      },
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [`"div"`]
      }
    ])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_1`
        }
      },
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          type: NodeTypes.SIMPLE_EXPRESSION,
          content: `_hoisted_2`
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist components', () => {
    const { root, args } = transformWithHoist(`<div><Comp/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [`_component_Comp`]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist element with dynamic props', () => {
    const { root, args } = transformWithHoist(`<div><div :id="foo"/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            createObjectMatcher({
              id: `[foo]`
            }),
            `null`,
            genFlagText(PatchFlags.PROPS),
            `["id"]`
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist element with static key', () => {
    const { root, args } = transformWithHoist(`<div><div key="foo"/></div>`)
    expect(root.hoists.length).toBe(1)
    expect(root.hoists).toMatchObject([
      {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_VNODE,
        arguments: [`"div"`, createObjectMatcher({ key: 'foo' })]
      }
    ])
    expect(args).toMatchObject([
      `"div"`,
      `null`,
      [
        {
          type: NodeTypes.ELEMENT,
          codegenNode: {
            type: NodeTypes.SIMPLE_EXPRESSION,
            content: `_hoisted_1`
          }
        }
      ]
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should NOT hoist element with dynamic key', () => {
    const { root, args } = transformWithHoist(`<div><div :key="foo"/></div>`)
    expect(root.hoists.length).toBe(0)
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            createObjectMatcher({
              key: `[foo]`
            })
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist static props for elements with directives', () => {
    const { root, args } = transformWithHoist(
      `<div><div id="foo" v-foo/></div>`
    )
    expect(root.hoists).toMatchObject([createObjectMatcher({ id: 'foo' })])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: APPLY_DIRECTIVES,
          arguments: [
            {
              callee: CREATE_VNODE,
              arguments: [
                `"div"`,
                {
                  type: NodeTypes.SIMPLE_EXPRESSION,
                  content: `_hoisted_1`
                },
                `null`,
                genFlagText(PatchFlags.NEED_PATCH)
              ]
            },
            {
              type: NodeTypes.JS_ARRAY_EXPRESSION
            }
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist static props for elements with dynamic text children', () => {
    const { root, args } = transformWithHoist(
      `<div><div id="foo">{{ hello }}</div></div>`
    )
    expect(root.hoists).toMatchObject([createObjectMatcher({ id: 'foo' })])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            { content: `_hoisted_1` },
            { type: NodeTypes.INTERPOLATION },
            genFlagText(PatchFlags.TEXT)
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('hoist static props for elements with unhoistable children', () => {
    const { root, args } = transformWithHoist(
      `<div><div id="foo"><Comp/></div></div>`
    )
    expect(root.hoists).toMatchObject([createObjectMatcher({ id: 'foo' })])
    expect(args[2]).toMatchObject([
      {
        type: NodeTypes.ELEMENT,
        codegenNode: {
          callee: CREATE_VNODE,
          arguments: [
            `"div"`,
            { content: `_hoisted_1` },
            [{ type: NodeTypes.ELEMENT, tag: `Comp` }]
          ]
        }
      }
    ])
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should hoist v-if props/children if static', () => {
    const { root, args } = transformWithHoist(
      `<div><div v-if="ok" id="foo"><span/></div></div>`
    )
    expect(root.hoists).toMatchObject([
      createObjectMatcher({
        key: `[0]`, // key injected by v-if branch
        id: 'foo'
      }),
      {
        callee: CREATE_VNODE,
        arguments: [`"span"`]
      }
    ])
    expect(args[2][0].codegenNode).toMatchObject({
      type: NodeTypes.JS_SEQUENCE_EXPRESSION,
      expressions: [
        { callee: OPEN_BLOCK },
        {
          type: NodeTypes.JS_CONDITIONAL_EXPRESSION,
          consequent: {
            // blocks should NOT be hoisted
            callee: CREATE_BLOCK,
            arguments: [
              `"div"`,
              { content: `_hoisted_1` },
              [
                {
                  codegenNode: { content: `_hoisted_2` }
                }
              ]
            ]
          }
        }
      ]
    })
    expect(generate(root).code).toMatchSnapshot()
  })

  test('should hoist v-for children if static', () => {
    const { root, args } = transformWithHoist(
      `<div><div v-for="i in list" id="foo"><span/></div></div>`
    )
    expect(root.hoists).toMatchObject([
      createObjectMatcher({
        id: 'foo'
      }),
      {
        callee: CREATE_VNODE,
        arguments: [`"span"`]
      }
    ])
    const forBlockCodegen = args[2][0].codegenNode
    expect(forBlockCodegen).toMatchObject({
      type: NodeTypes.JS_SEQUENCE_EXPRESSION,
      expressions: [
        { callee: OPEN_BLOCK },
        {
          callee: CREATE_BLOCK,
          arguments: [
            FRAGMENT,
            `null`,
            {
              type: NodeTypes.JS_CALL_EXPRESSION,
              callee: RENDER_LIST
            },
            genFlagText(PatchFlags.UNKEYED_FRAGMENT)
          ]
        }
      ]
    })
    const innerBlockCodegen =
      forBlockCodegen.expressions[1].arguments[2].arguments[1].returns
    expect(innerBlockCodegen).toMatchObject({
      type: NodeTypes.JS_SEQUENCE_EXPRESSION,
      expressions: [
        { callee: OPEN_BLOCK },
        {
          callee: CREATE_BLOCK,
          arguments: [
            `"div"`,
            { content: `_hoisted_1` },
            [
              {
                codegenNode: { content: `_hoisted_2` }
              }
            ]
          ]
        }
      ]
    })
    expect(generate(root).code).toMatchSnapshot()
  })
})
