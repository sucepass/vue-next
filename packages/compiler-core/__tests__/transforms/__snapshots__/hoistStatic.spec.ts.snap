// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: hoistStatic transform hoist element with static key 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"div\\", { key: \\"foo\\" })

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist nested static tree 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"p\\", null, [
  _createVNode(\\"span\\"),
  _createVNode(\\"span\\")
])

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist nested static tree with comments 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode
const _Comment = Vue.Comment

const _hoisted_1 = _createVNode(\\"div\\", null, [
  _createVNode(_Comment, null, \\"comment\\")
])

return function render() {
  with (this) {
    const { createVNode: _createVNode, Comment: _Comment, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist siblings with common non-hoistable parent 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"span\\")
const _hoisted_2 = _createVNode(\\"div\\")

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1,
      _hoisted_2
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist simple element 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = _createVNode(\\"span\\", { class: \\"inline\\" }, \\"hello\\")

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _hoisted_1
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist static props for elements with directives 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }

return function render() {
  with (this) {
    const { createVNode: _createVNode, applyDirectives: _applyDirectives, resolveDirective: _resolveDirective, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _directive_foo = _resolveDirective(\\"foo\\")
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _applyDirectives(_createVNode(\\"div\\", _hoisted_1, null, 32 /* NEED_PATCH */), [
        [_directive_foo]
      ])
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist static props for elements with dynamic text children 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }

return function render() {
  with (this) {
    const { toString: _toString, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", _hoisted_1, _toString(hello), 1 /* TEXT */)
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform hoist static props for elements with unhoistable children 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }

return function render() {
  with (this) {
    const { resolveComponent: _resolveComponent, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _component_Comp = _resolveComponent(\\"Comp\\")
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", _hoisted_1, [
        _createVNode(_component_Comp)
      ])
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist components 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { resolveComponent: _resolveComponent, createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    const _component_Comp = _resolveComponent(\\"Comp\\")
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(_component_Comp)
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist element with dynamic key 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", { key: foo })
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist element with dynamic props 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      _createVNode(\\"div\\", { id: foo }, null, 8 /* PROPS */, [\\"id\\"])
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should NOT hoist root node 1`] = `
"const _Vue = Vue

return function render() {
  with (this) {
    const { createVNode: _createVNode, createBlock: _createBlock, openBlock: _openBlock } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\"))
  }
}"
`;

exports[`compiler: hoistStatic transform should hoist v-for children if static 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode

const _hoisted_1 = { id: \\"foo\\" }
const _hoisted_2 = _createVNode(\\"span\\")

return function render() {
  with (this) {
    const { renderList: _renderList, openBlock: _openBlock, createBlock: _createBlock, Fragment: _Fragment, createVNode: _createVNode } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      (_openBlock(), _createBlock(_Fragment, null, _renderList(list, (i) => {
        return (_openBlock(), _createBlock(\\"div\\", _hoisted_1, [
          _hoisted_2
        ]))
      }), 128 /* UNKEYED_FRAGMENT */))
    ]))
  }
}"
`;

exports[`compiler: hoistStatic transform should hoist v-if props/children if static 1`] = `
"const _Vue = Vue
const _createVNode = Vue.createVNode
const _Comment = Vue.Comment

const _hoisted_1 = {
  key: 0,
  id: \\"foo\\"
}
const _hoisted_2 = _createVNode(\\"span\\")

return function render() {
  with (this) {
    const { openBlock: _openBlock, createVNode: _createVNode, createBlock: _createBlock, Comment: _Comment } = _Vue
    
    return (_openBlock(), _createBlock(\\"div\\", null, [
      (_openBlock(), ok
        ? _createBlock(\\"div\\", _hoisted_1, [
            _hoisted_2
          ])
        : _createBlock(_Comment))
    ]))
  }
}"
`;
